Manage numcalc fixes

SDX-LV opened the pull request:
Hi,
Did a bit of debugging to get the new version of NumCalc Manager to work on Windows.
Plus some more fixes that affect Linux as well.

Now it works on Windows 10 as it should.

f-brinkmann commented on mesh2hrtf/NumCalc/manage_numcalc.py:
```diff
@@ -341,9 +338,9 @@ def manage_numcalc(project_path=os.getcwd(), numcalc_path=None,
             if os.name == 'nt':  # Windows detected
                 # create a log file for all print-outs
-                LogFileHandle = open(f"NC{step}-{step}_log.txt", "w")
+                LogFileHandle = open(f"{cwd}\\NC{step}-{step}_log.txt", "w")
```
```suggestion
                LogFileHandle = open(os.path.join(cwd, "NC{step}-{step}_log.txt", "w")
```
would be more robust

SDX-LV replied:
Yes. I cheated here because this is explicitly "Windows NT" code :) 
but you are right. I included your suggestion.

f-brinkmann commented on mesh2hrtf/NumCalc/manage_numcalc.py:
```diff
@@ -418,16 +415,22 @@ def _raise_error(message, text_color, log_file, confirm_errors):
 def _print_message(message, text_color, log_file):
     """Print message to console and log file"""
 
+    if os.name == 'nt':  # Windows detected
+        text_color = ''  # color codes do not work as intended on Win10
```
I would be fine with completely removing color output in this case. What do you think?

SDX-LV replied:
I don't mind to dropping the color support. Only it may be possible to fix this, because the NumCalcManager.py and other scripts that I used previously worked nicely with colors in Windows command prompt, so perhaps there is some trick to fix it?
Up to you, but as it is it does not hurt on Windows and probably does work on Linux?

f-brinkmann replied:
Yes it works on Linux - I'm also fine with leaving is as it is.

SDX-LV commented on mesh2hrtf/NumCalc/manage_numcalc.py:
```diff
@@ -75,23 +75,23 @@ def manage_numcalc(project_path=os.getcwd(), numcalc_path=None,
     current_time = time.strftime("%Y_%m_%d_%H-%M-%S", time.localtime())
     log_file = os.path.join(
-        project_path, f"numcalc_manager_{current_time}.txt")
+        project_path, f"manage_numcalc_{current_time}.txt")
```
Did it. Hopefully did not break anything.

SDX-LV commented on mesh2hrtf/NumCalc/manage_numcalc.py:
```diff
@@ -123,11 +123,7 @@ def manage_numcalc(project_path=os.getcwd(), numcalc_path=None,
-        all_projects = [os.path.dirname(project_path)]
-        log_file = os.path.join(project_path, '..', log_file)
-    elif os.path.isfile(os.path.join(project_path, 'parameters.json')):
+    if os.path.isfile(os.path.join(project_path, 'parameters.json')):
```
Done.

SDX-LV commented on mesh2hrtf/NumCalc/manage_numcalc.py:
```diff
@@ -284,6 +277,10 @@ def manage_numcalc(project_path=os.getcwd(), numcalc_path=None,
                 text_color_red, log_file, confirm_errors)
 
+        # assure highest first if demanded
+        if starting_order != "low":
+            instances_to_run = np.flip(instances_to_run, axis=0)
```
This I don't understand - please explain in more detail of make the change directly.

f-brinkmann replied:
The section

```python
if ram_available < instances_to_run[-1, 3] * ram_safety_factor:
      _raise_error((
          f"Available RAM is {round(ram_available, 2)} GB, but frequency"
          f" step {int(instances_to_run[0, 1])} of source "
          f"{int(instances_to_run[0, 1])} requires "
          f"{round(instances_to_run[0, 3] * ram_safety_factor, 2)} GB."),
          text_color_red, log_file, confirm_errors)
```
compares the available against the maximally required RAM. If we flip `instances_to_run` after this section the frequency step with the highest RAM consumption is in `instances_to_tun[-1]` and not in instances_to_tun[0]. The sections seemed to be inconsistent before your changes already - good that we discovered this :)

SDX-LV replied:
should be fixed now.

f-brinkmann commented:
I also thought about tracking the available ram, but did not do it because it might be buggy if other system processes take RAM in the meantime. In addition the safety factor you use for tracking the RAM differs from the one that we already have as a parameter. If you revert those changes, the pull would be ready for merging and testing from my side.

f-brinkmann commented:
thx - I will merge and test either tomorrow or Wednesday...

SDX-LV commented:
Hi Fabian,

**I commented-out "ram_budget" optimization** so the logic is effectively back to the previous version. 

As "ram_budget" starts instances with almost no delay, it is not likely that available RAM will change much due to other processes running on a computer. Also "ram_budget" optimization still waits for the full `time.sleep(wait_time)` as soon as the excess RAM condition is no longer valid and it is possible to add big extra margins which will reduce the time gain, but will make the logic extremely robust (for example if "ram_budget" would count 2x estimated RAM for every instance, it will still win at least a minute on a system with plenty of RAM).

As for safety factor - I added extra multiplier on top of the existing safety factor, because I saw that the existing safety factor was underestimating the real RAM usage, especially for small RAM demands. That's why I also added a minimum threshold of 2GB - if estimate says instance will need less than 2GB, I still count 2GB towards the "ram_budget".

I understand that it may not be a good time to add even more features, but such optimization does provide several minutes shorter simulation time by eliminating most of the idling time - worth making it work :) 

Best regards,
Sergejs

f-brinkmann commented:
I see - how about making this a new Pull. But instead of introducing additional safety margins, I would be in favor of using the parameters we already have. We could think about increasing the default for the RAM safety factor to 1.1 and suggest the users to leave at least 2 GB RAM unused. We could make this easier by allowing negative numbers for the `max_ram_load` parameter, e.g., -2 would leave 2 GB free. We could also make -2 the default if we wanted.

SDX-LV commented:
Very good! 
Then we have Windows compatibility in place as well and tutorials should be in sync.
